keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -165,57 +201,340 @@ ensure all tenants have rights to the images that are already present on a host.
+- `AlwaysVerify` - all the image pulls need reverification.
+
+#### Writing to the Cache
+
+There are two different stages when the kubelet updates its cache:
+1. **Before the image pull**
+    - An image pull intent record is written befor the actual pull occurs.
+    - This is is to mitigate a situation where the kubelet crashes during the pull and an
+      image that gets pulled in the meantime later appears as if it were *preloaded*.
+2. **After the pull**
+    - Once we know the result of the pull operation, we either remove the image pull
+      intent record (=> failed pull attempt), or we write a new record of the image
+      being pulled successfully and remove the record of the pull intent.
+    - Without this, if kubelet failed during an image pull, an existing record would
+      already grant image access irregardless of this pull operation result for any next
+      pull attempts.
+
+The kubelet uses the following Golang interface to interact with the cache:
+```go
+// ImagePullCredentials wraps credentials that can be used to pull an image.
+// One of the attributes must always be non-nil, and the others must be nil.
+type ImagePullCredentials struct {
+    KubernetesSecret *imagemanager.ImagePullSecret
+    NodeWideCredentials *bool
+    Anonymous *bool
+}
+
+// ImagePullManager keeps the state of images that were pulled and which are
+// currently still being pulled.
+// It should keep an internal state of images currently being pulled by the kubelet
+// in order to determine whether to destroy a "pulling" record should an image
+// pull fail.
+type ImagePullManager interface {
+    // RecordPullIntent records an intent to pull an image.
+    // It returns a number of current pulls requesting the same image.
+    RecordPullIntent(image string) (uint23, error)
+    // RecordImagePulled writes a record of an image being successfully pulled
+    // with ImagePullCredentials.
+    //
+    // If `credentials` is nil, it means the image was pulled successfully, but
+    // a reverification should be required upon next pull request. This can be
+    // useful in cases where the kubelet crashed in the middle of the pull but
+    // the image appears on the node after the kubelet restarts.
+    RecordImagePulled(imageRef string, credentials *ImagePullCredentials) error
+    // RecordImagePullFailed should be called if an image failed to pull.
+    RecordImagePullFailed(image string) error
+    // ShouldReverifyCredentials evaluates the policy for image specified in
+    // "image" and if the policy demands reverification, it checks the internal
+    // cache to see if there's a record of pulling the image with the presented
+    // set of credentials.
+    // Returns true if policy demands reverification and no record of the pull
+    // was found in the cache.
+    ShouldReverifyCredentials(image string, credentials *ImagePullCredentials) bool
```

Comment:
maybe we want a different arg to ShouldReverify so we can make a single call instead of in a loop (and *shouldreverify* doesn't really need to explicitly opt into checking anonymous / node-wide)

Reply:
doc the `image` arg corresponds to the pod spec

if we only call this when we've already resolved `image` → and existing `imageRef` on the node, should we pass that info in to this call so the implementation doesn't have to resolve it again?

Reply:
In the past implementation, we make a *credentialprovider.DockerKeyring from the pullSecrets.

https://github.com/kubernetes/kubernetes/blob/21f54c3ca1ac87611e101f46d8b05682f51efbee/pkg/kubelet/kuberuntime/kuberuntime_image.go#L42-L82